src/SArray.jl

```diff
@@ -50,6 +50,34 @@ end
     end
 end
 
+@generated function (::Type{SArray{S, T}})(gen::Base.Generator) where {S <: Tuple, T}
```

Comment:
This method is useful for arbitrary iterators. Maybe we can have `sacollect` or something?

(I wish we had something like `Base.into(ContainerType, itr)`...)

Reply:
`sacollect` is a good idea. How would it look like? Should it take a `Val` argument to specify the vector length?

Reply:
Since we already have `SArray{S}`, `SVector{L}`, etc. as an interface for defining size, I think it'd be nice to use it as the interface for collect-like function.  We also have `Size` for this but then we'd need a separate mechanism for specifying the container type (e.g., `MVector`, `FieldVector`, ...).

Maybe the best action here is to define (say) `collect_to(::Type{<:StaticArray{S, T}}, itr)` as an internal function and then separately discuss how to expose it as an API?

Reply:
It's going to take the expected result type as first argument.

Reply:
It's currently called `sacollect(::Type{SA}, gen) where {SA <: SArray{S}} where {S <: Tuple}`, and otherwise identical to your `collect_to`. It's also exported.

What do you want to see in this PR? Rename to `collect_to` and not exported?

Reply:
I suggested the name `sacollect` but then I didn't want it to be a public API because it's a bit different from `collect`.  The first argument of `collect(::Type, collection)` is an *element* type while `sacollect(::Type, collection)` ATM here is a *storage* type.  I think it'd be nice to come up with the API that does not conflict with the one in `Base`.

The only function in `Base` that I know and is close to `sacollect` here is the internal function `Base.collect_to!(dest, ...)` although the first argument is an array, not a type.  That's why I suggested `StaticArrays.collect_to(::Type, collection)`.

I think the cleanest solution would be to define something in `Base` (e.g., `into(output_storage_type, input_collection)`) but of course that would take some time and by no means I want to block this PR.  I think @andyferris and @c42f might want to look into this.

`sacollect` may be a good short/mid-term solution. If other people are OK with this I'm not stopping it.

Reply:
How should this be resolved?

Since there isn't a good way to do this in a `Base`-compatible way, I suggest to keep `sacollect` since this provides a functionality that is not available otherwise.

Reply:
I'm happy to merge this now if `sacollect` is not exported. We can discuss the way to export it in a different issue.

Reply:
I made sacollect private.